File: packages/clippster-editor-native/android/src/main/java/app/clippster/editor/engine/ClippsterEditorNativeModule.kt
package app.clippster.editor.engine

import android.media.MediaExtractor
import android.media.MediaFormat
import android.media.MediaMetadataRetriever
import android.net.Uri
import android.os.Handler
import android.os.Looper
import android.view.Choreographer
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.io.FileOutputStream
import java.util.concurrent.atomic.AtomicReference

class ClippsterEditorNativeModule : Module() {
  private val mainHandler = Handler(Looper.getMainLooper())
  private val exporter = HardwareExportPipeline()
  private val document = AtomicReference("")

  @Volatile private var playing = false
  @Volatile private var previewQuality = "auto"
  @Volatile private var currentSeconds = 0.0
  private var playStartedNanos = 0L
  private var playStartedSeconds = 0.0

  private val frameCallback = object : Choreographer.FrameCallback {
    override fun doFrame(frameTimeNanos: Long) {
      if (!playing) return
      currentSeconds = playStartedSeconds +
        (frameTimeNanos - playStartedNanos).coerceAtLeast(0L) / 1_000_000_000.0
      sendEvent("onTimeUpdate", mapOf("timeSeconds" to currentSeconds))
      Choreographer.getInstance().postFrameCallback(this)
    }
  }

  override fun definition() = ModuleDefinition {
    Name("ClippsterEditorNative")

    Constants(
      "ticksPerSecond" to nativeTicksPerSecond(),
    )

    Events(
      "onTimeUpdate",
      "onExportProgress",
      "onExportComplete",
      "onExportError",
      "onEngineError",
    )

    AsyncFunction("getCapabilities") {
      jsonArrayToList(JSONArray(nativeGetCapabilitiesJson()))
    }

    AsyncFunction("loadRevision") { documentJson: String ->
      validateDocument(documentJson)
      document.set(documentJson)
      Unit
    }

    AsyncFunction("applyRevision") { documentJson: String ->
      validateDocument(documentJson)
      document.set(documentJson)
      Unit
    }

    AsyncFunction("play") {
      if (!playing) {
        playing = true
        playStartedSeconds = currentSeconds
        mainHandler.post {
          playStartedNanos = System.nanoTime()
          Choreographer.getInstance().removeFrameCallback(frameCallback)
          Choreographer.getInstance().postFrameCallback(frameCallback)
        }
      }
      Unit
    }

    AsyncFunction("pause") {
      playing = false
      mainHandler.post { Choreographer.getInstance().removeFrameCallback(frameCallback) }
      Unit
    }

    AsyncFunction("seek") { timeSeconds: Double, mode: String ->
      require(timeSeconds.isFinite() && timeSeconds >= 0.0) { "Invalid seek time" }
      require(mode == "interactive" || mode == "precise") {
        "Seek mode must be interactive or precise"
      }
      currentSeconds = timeSeconds
      playStartedSeconds = timeSeconds
      playStartedNanos = System.nanoTime()
      sendEvent("onTimeUpdate", mapOf("timeSeconds" to currentSeconds, "mode" to mode))
      Unit
    }

    AsyncFunction("setPreviewQuality") { quality: String ->
      require(quality in setOf("auto", "low", "medium", "high", "full")) {
        "Unsupported preview quality: $quality"
      }
      previewQuality = quality
      Unit
    }

    AsyncFunction("getCurrentTime") {
      currentSeconds
    }

    AsyncFunction("export") { requestJson: String ->
      val request = JSONObject(requestJson)
      val outputPath = request.getString("outputPath")
      try {
        val scenePayload = when {
          request.has("sceneJson") && !request.isNull("sceneJson") ->
            request.getString("sceneJson")
          request.has("documentJson") && !request.isNull("documentJson") ->
            request.getString("documentJson")
          else -> null
        }
        val outputPaths = if (scenePayload != null) {
          exporter.exportScene(
            context = requireNotNull(appContext.reactContext),
            sceneJson = scenePayload,
            outputPath = outputPath,
            width = request.optInt("width", 1080),
            height = request.optInt("height", 1920),
            fps = request.optInt("fps", 30),
          ) { progress ->
            sendEvent("onExportProgress", mapOf("progress" to progress))
          }
        } else {
          require(request.has("inputPath") && !request.isNull("inputPath")) {
            "Export requires sceneJson, documentJson, or inputPath"
          }
          exporter.export(
            inputPath = request.getString("inputPath"),
            outputPath = outputPath,
          ) { progress ->
            sendEvent("onExportProgress", mapOf("progress" to progress))
          }
        }
        val result = mapOf("outputPaths" to outputPaths)
        sendEvent("onExportComplete", result)
        result
      } catch (error: Throwable) {
        sendEvent("onExportError", mapOf("message" to (error.message ?: "Export failed")))
        throw error
      }
    }

    AsyncFunction("cancelExport") {
      exporter.cancel()
      Unit
    }

    AsyncFunction("probeMedia") { sourceUri: String ->
      probeMediaMetadata(sourceUri)
    }

    AsyncFunction("generateProxy") { sourceUri: String, destUri: String ->
      val source = fileForUri(sourceUri)
      val destination = fileForUri(destUri)
      destination.parentFile?.mkdirs()
      source.copyTo(destination, overwrite = true)
      destination.absolutePath
    }

    AsyncFunction("generateThumbnail") {
        sourceUri: String, timeSeconds: Double, destUri: String ->
      val retriever = MediaMetadataRetriever()
      try {
        val source = Uri.parse(sourceUri)
        val context = requireNotNull(appContext.reactContext)
        if (source.scheme == "content" || source.scheme == "file") {
          retriever.setDataSource(context, source)
        } else {
          retriever.setDataSource(sourceUri)
        }
        val bitmap = retriever.getFrameAtTime(
          (timeSeconds.coerceAtLeast(0.0) * 1_000_000.0).toLong(),
          MediaMetadataRetriever.OPTION_CLOSEST_SYNC,
        ) ?: error("No thumbnail frame available")
        val destination = fileForUri(destUri)
        destination.parentFile?.mkdirs()
        FileOutputStream(destination).use {
          check(bitmap.compress(android.graphics.Bitmap.CompressFormat.JPEG, 90, it)) {
            "Could not encode thumbnail"
          }
        }
        bitmap.recycle()
        destination.absolutePath
      } finally {
        retriever.release()
      }
    }

    View(ClippsterEditorPreviewView::class) {
      Events("onSurfaceReady", "onFramePresented")
      Prop("documentJson") { view: ClippsterEditorPreviewView, value: String ->
        view.setDocumentJson(value)
      }
      Prop("playing") { view: ClippsterEditorPreviewView, value: Boolean ->
        view.setPlaying(value)
      }
      Prop("playheadSeconds") { view: ClippsterEditorPreviewView, value: Double ->
        view.setPlayheadSeconds(value)
      }
      Prop("quality") { view: ClippsterEditorPreviewView, value: String ->
        view.setQuality(value)
      }
    }
  }

  private fun validateDocument(value: String) {
    require(value.isNotBlank()) { "Document JSON is empty" }
    val tick = (currentSeconds * nativeTicksPerSecond()).toLong()
    val result = JSONObject(nativeParseAndEvaluate(value, tick, true))
    if (result.has("error")) {
      val message = result.optString("error", "Invalid editor document")
      sendEvent("onEngineError", mapOf("message" to message))
      error(message)
    }
  }

  private fun fileForUri(value: String): File {
    val uri = Uri.parse(value)
    require(uri.scheme == null || uri.scheme == "file") {
      "Destination and copy paths must be local file URIs"
    }
    return File(if (uri.scheme == "file") requireNotNull(uri.path) else value)
  }

  private fun probeMediaMetadata(sourceUri: String): Map<String, Any?> {
    val path = fileForUri(sourceUri).absolutePath
    require(File(path).isFile) { "Media file does not exist: $path" }

    val retriever = MediaMetadataRetriever()
    val width: Int
    val height: Int
    val durationSeconds: Double
    try {
      retriever.setDataSource(path)
      width = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_VIDEO_WIDTH)
        ?.toIntOrNull()
        ?: error("Could not read export width")
      height = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_VIDEO_HEIGHT)
        ?.toIntOrNull()
        ?: error("Could not read export height")
      val durationMs = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION)
        ?.toLongOrNull()
        ?: error("Could not read export duration")
      durationSeconds = durationMs / 1000.0
    } finally {
      retriever.release()
    }

    var videoCodec = "h264"
    var audioCodec: String? = null
    val extractor = MediaExtractor()
    try {
      extractor.setDataSource(path)
      for (index in 0 until extractor.trackCount) {
        val format = extractor.getTrackFormat(index)
        val mime = format.getString(MediaFormat.KEY_MIME) ?: continue
        when {
          mime.startsWith("video/") ->
            videoCodec = when {
              mime.contains("avc", ignoreCase = true) || mime.contains("h264", ignoreCase = true) ->
                "h264"
              mime.contains("hevc", ignoreCase = true) || mime.contains("h265", ignoreCase = true) ->
                "hevc"
              else -> mime.substringAfter('/')
            }
          mime.startsWith("audio/") ->
            audioCodec = when {
              mime.contains("mp4a", ignoreCase = true) || mime.contains("aac", ignoreCase = true) ->
                "aac"
              else -> mime.substringAfter('/')
            }
        }
      }
    } finally {
      extractor.release()
    }

    return mapOf(
      "width" to width,
      "height" to height,
      "duration" to durationSeconds,
      "videoCodec" to videoCodec,
      "audioCodec" to audioCodec,
    )
  }

  private fun jsonArrayToList(array: JSONArray): List<Map<String, Any?>> =
    (0 until array.length()).map { jsonObjectToMap(array.getJSONObject(it)) }

  private fun jsonObjectToMap(value: JSONObject): Map<String, Any?> =
    value.keys().asSequence().associateWith { key ->
      when (val entry = value.get(key)) {
        JSONObject.NULL -> null
        is JSONObject -> jsonObjectToMap(entry)
        is JSONArray -> (0 until entry.length()).map(entry::get)
        else -> entry
      }
    }

  internal companion object {
    init {
      System.loadLibrary("clippster_editor_jni")
    }

    @JvmStatic private external fun nativeGetCapabilitiesJson(): String
    @JvmStatic private external fun nativeParseAndEvaluate(
      sceneJson: String,
      tick: Long,
      previewMode: Boolean,
    ): String
    @JvmStatic private external fun nativeTicksPerSecond(): Long

    fun evaluateDocument(sceneJson: String, tick: Long, previewMode: Boolean): String =
      nativeParseAndEvaluate(sceneJson, tick, previewMode)

    fun ticksPerSecond(): Long = nativeTicksPerSecond()
  }
}

File: packages/clippster-editor-native/ios/ClippsterEditorNativeModule.swift
import AVFoundation
import ExpoModulesCore
import Foundation
import ImageIO

public final class ClippsterEditorNativeModule: Module {
  private let exporter = HardwareExportPipeline()
  private let stateQueue = DispatchQueue(label: "app.clippster.editor.state")
  private var documentJSON = ""
  private var isPlaying = false
  private var previewQuality = "auto"
  private var currentSeconds = 0.0
  private var playStartSeconds = 0.0
  private var playStartUptime = ProcessInfo.processInfo.systemUptime
  private var timer: DispatchSourceTimer?

  public func definition() -> ModuleDefinition {
    Name("ClippsterEditorNative")

    Constants([
      "ticksPerSecond": GraphBridge.ticksPerSecond(),
    ])

    Events(
      "onTimeUpdate",
      "onExportProgress",
      "onExportComplete",
      "onExportError",
      "onEngineError"
    )

    AsyncFunction("getCapabilities") { () -> [[String: Any]] in
      let data = Data(GraphBridge.capabilitiesJSON().utf8)
      return try JSONSerialization.jsonObject(with: data) as? [[String: Any]] ?? []
    }

    AsyncFunction("loadRevision") { (value: String) in
      try self.validateDocument(value)
      self.stateQueue.sync { self.documentJSON = value }
    }

    AsyncFunction("applyRevision") { (value: String) in
      try self.validateDocument(value)
      self.stateQueue.sync { self.documentJSON = value }
    }

    AsyncFunction("play") {
      self.stateQueue.sync {
        guard !self.isPlaying else { return }
        self.isPlaying = true
        self.playStartSeconds = self.currentSeconds
        self.playStartUptime = ProcessInfo.processInfo.systemUptime
      }
      self.startClock()
    }

    AsyncFunction("pause") {
      self.stateQueue.sync { self.isPlaying = false }
      self.stopClock()
    }

    AsyncFunction("seek") { (timeSeconds: Double, mode: String) in
      guard timeSeconds.isFinite, timeSeconds >= 0 else {
        throw ModuleError.invalidArgument("Invalid seek time")
      }
      guard mode == "interactive" || mode == "precise" else {
        throw ModuleError.invalidArgument("Seek mode must be interactive or precise")
      }
      self.stateQueue.sync {
        self.currentSeconds = timeSeconds
        self.playStartSeconds = timeSeconds
        self.playStartUptime = ProcessInfo.processInfo.systemUptime
      }
      self.sendEvent("onTimeUpdate", ["timeSeconds": timeSeconds, "mode": mode])
    }

    AsyncFunction("setPreviewQuality") { (quality: String) in
      guard ["auto", "low", "medium", "high", "full"].contains(quality) else {
        throw ModuleError.invalidArgument("Unsupported preview quality: \(quality)")
      }
      self.stateQueue.sync { self.previewQuality = quality }
    }

    AsyncFunction("getCurrentTime") { () -> Double in
      self.stateQueue.sync { self.currentSeconds }
    }

    AsyncFunction("export") { (requestJSON: String) -> [String: Any] in
      do {
        guard
          let data = requestJSON.data(using: .utf8),
          let request = try JSONSerialization.jsonObject(with: data) as? [String: Any],
          let outputPath = request["outputPath"] as? String
        else { throw ModuleError.invalidArgument("Invalid export request JSON") }

        let outputPaths: [String]
        if let sceneJSON = request["sceneJson"] as? String {
          outputPaths = try self.exporter.exportScene(
            sceneJSON: sceneJSON,
            outputPath: outputPath,
            width: request["width"] as? Int ?? 1080,
            height: request["height"] as? Int ?? 1920,
            fps: request["fps"] as? Int ?? 30
          ) { progress in
            self.sendEvent("onExportProgress", ["progress": progress])
          }
        } else if let inputPath = request["inputPath"] as? String {
          outputPaths = try self.exporter.export(
            inputPath: inputPath,
            outputPath: outputPath
          ) { progress in
            self.sendEvent("onExportProgress", ["progress": progress])
          }
        } else {
          throw ModuleError.invalidArgument("Export request needs inputPath or sceneJson")
        }
        let result: [String: Any] = ["outputPaths": outputPaths]
        self.sendEvent("onExportComplete", result)
        return result
      } catch {
        self.sendEvent("onExportError", ["message": error.localizedDescription])
        throw error
      }
    }

    AsyncFunction("cancelExport") {
      self.exporter.cancel()
    }

    AsyncFunction("probeMedia") { (sourceURI: String) -> [String: Any?] in
      try Self.probeMediaMetadata(sourceURI)
    }

    AsyncFunction("generateProxy") {
      (sourceURI: String, destinationURI: String) -> [String: Any] in
      let source = Self.fileURL(sourceURI)
      let destination = Self.fileURL(destinationURI)
      try FileManager.default.createDirectory(
        at: destination.deletingLastPathComponent(),
        withIntermediateDirectories: true
      )
      try? FileManager.default.removeItem(at: destination)
      try FileManager.default.copyItem(at: source, to: destination)
      return [
        "outputPath": destination.path,
        "passthrough": true,
        "note": "Proxy transcoding is not implemented; source was copied.",
      ]
    }

    AsyncFunction("generateThumbnail") {
      (sourceURI: String, timeSeconds: Double, destinationURI: String) -> [String: Any] in
      let asset = AVURLAsset(url: Self.fileURL(sourceURI))
      let generator = AVAssetImageGenerator(asset: asset)
      generator.appliesPreferredTrackTransform = true
      let image = try generator.copyCGImage(
        at: CMTime(seconds: max(0, timeSeconds), preferredTimescale: 600),
        actualTime: nil
      )
      let destination = Self.fileURL(destinationURI)
      try FileManager.default.createDirectory(
        at: destination.deletingLastPathComponent(),
        withIntermediateDirectories: true
      )
      guard
        let target = CGImageDestinationCreateWithURL(
          destination as CFURL, "public.jpeg" as CFString, 1, nil
        )
      else { throw ModuleError.invalidArgument("Could not create thumbnail destination") }
      CGImageDestinationAddImage(target, image, [
        kCGImageDestinationLossyCompressionQuality: 0.9,
      ] as CFDictionary)
      guard CGImageDestinationFinalize(target) else {
        throw ModuleError.invalidArgument("Could not encode thumbnail")
      }
      return ["outputPath": destination.path]
    }

    View(ClippsterEditorPreviewView.self) {
      Events("onSurfaceReady", "onFramePresented")
      Prop("documentJson") { (view, value: String) in view.setDocumentJSON(value) }
      Prop("playing") { (view, value: Bool) in view.setPlaying(value) }
      Prop("playheadSeconds") { (view, value: Double) in view.setPlayheadSeconds(value) }
      Prop("quality") { (view, value: String) in view.setQuality(value) }
    }

    OnDestroy {
      self.stopClock()
      self.exporter.cancel()
    }
  }

  private func validateDocument(_ value: String) throws {
    guard !value.isEmpty else { throw ModuleError.invalidArgument("Document JSON is empty") }
    let tick = stateQueue.sync {
      Int64(currentSeconds * Double(GraphBridge.ticksPerSecond()))
    }
    do {
      _ = try GraphBridge.parseAndEvaluate(value, tick: tick, previewMode: true)
    } catch {
      sendEvent("onEngineError", ["message": error.localizedDescription])
      throw error
    }
  }

  private func startClock() {
    DispatchQueue.main.async {
      guard self.timer == nil else { return }
      let timer = DispatchSource.makeTimerSource(queue: .main)
      timer.schedule(deadline: .now(), repeating: 1.0 / 30.0)
      timer.setEventHandler { [weak self] in
        guard let self else { return }
        let value = self.stateQueue.sync { () -> Double in
          guard self.isPlaying else { return self.currentSeconds }
          self.currentSeconds = self.playStartSeconds
            + ProcessInfo.processInfo.systemUptime - self.playStartUptime
          return self.currentSeconds
        }
        self.sendEvent("onTimeUpdate", ["timeSeconds": value])
      }
      self.timer = timer
      timer.resume()
    }
  }

  private func stopClock() {
    DispatchQueue.main.async {
      self.timer?.cancel()
      self.timer = nil
    }
  }

  private static func fileURL(_ value: String) -> URL {
    if let url = URL(string: value), url.isFileURL { return url }
    return URL(fileURLWithPath: value)
  }

  private static func probeMediaMetadata(_ sourceURI: String) throws -> [String: Any?] {
    let url = fileURL(sourceURI)
    guard FileManager.default.fileExists(atPath: url.path) else {
      throw ModuleError.invalidArgument("Media file does not exist: \(url.path)")
    }
    let asset = AVURLAsset(url: url)
    let videoTrack = asset.tracks(withMediaType: .video).first
    let audioTrack = asset.tracks(withMediaType: .audio).first
    guard let videoTrack else {
      throw ModuleError.invalidArgument("Could not validate exported video metadata")
    }
    let size = videoTrack.naturalSize.applying(videoTrack.preferredTransform)
    let width = Int(abs(size.width).rounded())
    let height = Int(abs(size.height).rounded())
    let duration = CMTimeGetSeconds(asset.duration)
    guard width > 0, height > 0, duration.isFinite, duration > 0 else {
      throw ModuleError.invalidArgument("Could not validate exported video metadata")
    }
    let videoCodec: String
    if let format = videoTrack.formatDescriptions.first {
      let type = CMFormatDescriptionGetMediaSubType(format as! CMFormatDescription)
      videoCodec = type == kCMVideoCodecType_H264 ? "h264" : String(format: "%c%c%c%c",
        (type >> 24) & 0xff, (type >> 16) & 0xff, (type >> 8) & 0xff, type & 0xff)
    } else {
      videoCodec = "h264"
    }
    var audioCodec: String? = nil
    if let audioTrack,
       let format = audioTrack.formatDescriptions.first {
      let type = CMFormatDescriptionGetMediaSubType(format as! CMFormatDescription)
      audioCodec = type == kAudioFormatMPEG4AAC ? "aac" : String(format: "%c%c%c%c",
        (type >> 24) & 0xff, (type >> 16) & 0xff, (type >> 8) & 0xff, type & 0xff)
    }
    return [
      "width": width,
      "height": height,
      "duration": duration,
      "videoCodec": videoCodec.lowercased(),
      "audioCodec": audioCodec?.lowercased(),
    ]
  }

  private enum ModuleError: LocalizedError {
    case invalidArgument(String)

    var errorDescription: String? {
      switch self {
      case .invalidArgument(let message): return message
      }
    }
  }
}
